ospfd/ospf_flood.h

```diff
@@ -7,6 +7,19 @@
 #ifndef _ZEBRA_OSPF_FLOOD_H
 #define _ZEBRA_OSPF_FLOOD_H
 
+/*
+ * OSPF Temporal list - Used for Link State Retransmission List Nodes.
+ */
+struct ospf_temporal_listnode {
+	struct listnode listnode;
```

Comment:
A general comment: we'd really like to encourage use of the typesafe collections, and avoid new use of the various older collection apis. since this is a brand-new object, could it be converted to use the typesafe list?

Reply:
It looks to me like this requires a DECLARE_DLIST() element that doesn't contain anything but an item pointer? This would require a two pieces of memory for each list? Also, it is not apparent from typesafe.h which are the internal and which are the external APIs. This looks to me like an exercise obfuscation. Conversely, the linklist.h defines and APIs are very understandable (although the usage of listset_app_node_mem() could be better explained). Also, all the required APIs are provided (including listnode_move_to_tail()). 

Can you tell me how'd I implement this without having multiple list items for each list element?

Reply:
the benefit of the typesafe lib is that it's ... strongly-typed; it also produces in per-collection apis, so that makes things clear (at least in the general community).
the internal implementation isn't really exotic: there's an internal linkage node in the definition of the type that's to be on the list, and that's it? I didn't understand "multiple list items for each list element" - what's the issue there?

Reply:
I read it as there are going to be multiple lists all pointing at the same data structure.  This is possible to do with the typesafe but it's a bit more work.  

I think it is a reasonable request to add a move_to_tail to the typesafe structures for non-ordered lists.

Reply:
> I read it as there are going to be multiple lists all pointing at the same data structure. This is possible to do with the typesafe but it's a bit more work.
> 
> I think it is a reasonable request to add a move_to_tail to the typesafe structures for non-ordered lists.

Is it possible to imbed the dlist next/prev structures in another struct and only have one level of list node?  Is there an example of this anywhere? Maybe It is apparent to everyone else but typesafe.[c,h] are hard to reverse engineer and, at least on the surface, I can't see any advantage.

Reply:
I'm still not really tracking the problem I guess.
There's an example of using an intermediate type as linkage (in rbtrees) in zebra/zebra_nhg.h and zebra/zebra_nhg_private.h
 But if you want to have two different lists, then the objects involved need two connection nodes, that is true. I think that's a benefit - you can't use the wrong api with the wrong internal node, the two lists "know" what they're allowed to touch inside the object, and they won't interfere with one another.

Reply:
The problem is that I only want one exogenous list element, i.e., {timestamp, list-linkage, lsa pointer}. I don't want two exogenous items per list element {list-linkage item pointer}-> {timestamp, lsa pointer}. 

If the dlist in typesafe.[c,h] is easy to use, someone should be able to easily tell me how to do it...

Reply:
This documentation is https://docs.frrouting.org/projects/dev-guide/en/latest/lists.html is ok but could benefit from pointing to examples in FRR files. Of course, these aren't too hard to find once you've decided on a type safe data structure.